File: android/app/src/main/java/com/dismoi/scout/Floating/FloatingModule.java
package com.dismoi.scout.Floating;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageButton;

import androidx.annotation.NonNull;

import com.dismoi.scout.Floating.Layout.Bubble;
import com.dismoi.scout.Floating.Layout.Message;
import com.dismoi.scout.Floating.OnCallback;
import com.dismoi.scout.R;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import android.annotation.SuppressLint;

public class FloatingModule extends ReactContextBaseJavaModule {

  private Manager bubblesManager;
  private Manager messagesManager;
  private final ReactApplicationContext reactContext;
  private Bubble bubbleDisMoiView;
  private Message messageDisMoiView;

  public FloatingModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  @ReactMethod
  public void reopenApp(){
    Intent launchIntent = reactContext.getPackageManager().getLaunchIntentForPackage(reactContext.getPackageName());
    if (launchIntent != null) {
      reactContext.startActivity(launchIntent);
    }
  }

  @NonNull
  @Override
  public String getName() {
    return "FloatingModule";
  }

  @ReactMethod
  public void showFloatingDisMoiBubble(int x, int y, final Promise promise) {
    try {
      this.addNewFloatingDisMoiBubble(x, y);
      promise.resolve("");
    } catch (Exception e) {
      promise.reject("");
    }
  }

  @ReactMethod
  public void showFloatingDisMoiMessage(int x, int y, final Promise promise) {
    if (messageDisMoiView == null) {
      try {
        this.addNewFloatingDisMoiMessage(x, y);
        promise.resolve("");
      } catch (Exception e) {
        promise.reject("");
      }
    }
  }

  @ReactMethod
  public void hideFloatingDisMoiBubble(final Promise promise) {
    try {
      this.removeDisMoiBubble();
      promise.resolve("");
    } catch (Exception e) {
      promise.reject("");
    }
  }

  @ReactMethod
  public void hideFloatingDisMoiMessage(final Promise promise) {
    try {
      this.removeDisMoiMessage();
      promise.resolve("");
    } catch (Exception e) {
      promise.reject("");
    }
  }
  
  @ReactMethod
  public void requestPermission(final Promise promise) {
    this.requestPermissionAction(promise);
  }  
  
  @ReactMethod
  public void checkPermission(final Promise promise) {
    try {
      promise.resolve(hasPermission());
    } catch (Exception e) {
      promise.reject("");
    }
  }  
  
  @ReactMethod
  public void initialize(final Promise promise) {
    try {
      this.initializeBubblesManager();
      this.initializeDisMoiMessageManager();
      promise.resolve("");
    } catch (Exception e) {
      promise.reject("");
    }
  }

  private void addNewFloatingDisMoiMessage(int x, int y) {
    this.removeDisMoiBubble();

    messageDisMoiView = (Message) LayoutInflater.from(reactContext).inflate(R.layout.dismoi_message, messageDisMoiView, false);

    ImageButton imageButton = (ImageButton) messageDisMoiView.findViewById(R.id.close);
    imageButton.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        sendEventToReactNative("floating-dismoi-message-press");
      }
    });

    messagesManager.addDisMoiMessage(messageDisMoiView, x, y);
  }

  private void addNewFloatingDisMoiBubble(int x, int y) {

    bubbleDisMoiView = (Bubble) LayoutInflater.from(reactContext).inflate(R.layout.bubble_layout, bubbleDisMoiView, false);

    bubbleDisMoiView.setOnBubbleRemoveListener(new Bubble.OnBubbleRemoveListener() {
      @Override
      public void onBubbleRemoved(Bubble bubble) {
        bubbleDisMoiView = null;
        sendEventToReactNative("floating-dismoi-bubble-remove");
      }
    });
    bubbleDisMoiView.setOnBubbleClickListener(new Bubble.OnBubbleClickListener() {
      @Override
      public void onBubbleClick(Bubble bubble) {
        sendEventToReactNative("floating-dismoi-bubble-press");
      }
    });
    bubbleDisMoiView.setShouldStickToWall(true);
    bubblesManager.addDisMoiBubble(bubbleDisMoiView, x, y);
  }

  private boolean hasPermission(){
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      return Settings.canDrawOverlays(reactContext);
    }
    return true;
  }

  private void removeDisMoiBubble() {
    if (bubbleDisMoiView != null) {
      bubblesManager.removeBubble(bubbleDisMoiView);
    }
  }

  private void removeDisMoiMessage() {
    if (messageDisMoiView != null) {
      messagesManager.removeDisMoiMessage(messageDisMoiView);
    }
  }

  public void requestPermissionAction(final Promise promise) {
    if (!hasPermission()) {
      Intent intent = null;
      if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
        intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:" + reactContext.getPackageName()));
      }
      Bundle bundle = new Bundle();
      reactContext.startActivityForResult(intent, 0, bundle);
    } 
    if (hasPermission()) {
      promise.resolve("");
    } else {
      promise.reject("");
    }
  }

  private void initializeDisMoiMessageManager() {

    messagesManager = new Manager.Builder(reactContext).setTrashLayout(R.layout.bubble_trash_layout).setInitializationCallback(new OnCallback() {
      @Override
      public void onInitialized() {}
    }).build();

    messagesManager.initialize();
  }

  private void initializeBubblesManager() {

    bubblesManager = new Manager.Builder(reactContext).setTrashLayout(R.layout.bubble_trash_layout).setInitializationCallback(new OnCallback() {
      @Override
      public void onInitialized() {}
    }).build();

    bubblesManager.initialize();
  }

  private void sendEventToReactNative(String eventName) {
    WritableMap params = Arguments.createMap();
    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
      .emit(eventName, params);
  }
}

File: android/app/src/main/java/com/dismoi/scout/floating/FloatingModule.kt
package com.dismoi.scout.floating

import android.net.Uri
import android.os.Build
import android.text.Html
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.style.URLSpan
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import com.dismoi.scout.R
import com.dismoi.scout.floating.layout.Bubble
import com.dismoi.scout.floating.layout.Bubble.OnBubbleClickListener
import com.dismoi.scout.floating.layout.Bubble.OnBubbleRemoveListener
import com.dismoi.scout.floating.layout.Message
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import com.synnapps.carouselview.CarouselView
import com.synnapps.carouselview.ViewListener
import com.google.android.material.bottomsheet.BottomSheetBehavior

class FloatingModule(
  private val reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext) {

  var bubbleIsInitialized = false
  var messageIsInitialized = false

  private var bubblesManager: Manager? = Manager.Builder(reactContext).setTrashLayout(
    R.layout.bubble_trash
  ).setInitializationCallback(object : OnCallback {
    override fun onInitialized() {
      Log.d("Notification", "Bubble Is initialized")
      bubbleIsInitialized = true
    }
  }).build()

  private var messagesManager: Manager? = Manager.Builder(reactContext)
  .setInitializationCallback(object : OnCallback {
    override fun onInitialized() {
      Log.d("Notification", "Message Is initialized")
      messageIsInitialized = true
    }
  }).build()

  private var bubbleDisMoiView: Bubble? = null
  private var messageDisMoiView: Message? = null
  private var _size = 0
  private lateinit var _notices: ReadableArray
  private var _url = ""

  var verticalLayout: LinearLayout? = null

  @ReactMethod
  fun reopenApp() {
    val launchIntent = reactContext.packageManager.getLaunchIntentForPackage(
      reactContext.packageName
    )
    if (launchIntent != null) {
      reactContext.startActivity(launchIntent)
    }
  }

  override fun getName(): String {
    return "FloatingModule"
  }

  @ReactMethod
  fun showFloatingDisMoiBubble(
    x: Int,
    y: Int,
    numberOfNotice: Int,
    notices: ReadableArray,
    url: String,
    promise: Promise
  ) {
    bubblesManager!!.initialize()

    if (bubbleDisMoiView == null && messageDisMoiView == null) {
      _url = url
      _notices = notices
      _size = numberOfNotice
      addNewFloatingDisMoiBubble(x, y, numberOfNotice.toString())
      promise.resolve("")
    }
  }

  @ReactMethod
  fun showFloatingDisMoiMessage(x: Int, y: Int, promise: Promise) {
    bubblesManager!!.initialize()
    messagesManager!!.initialize()

    if (messageDisMoiView == null) {
      try {
        if (bubbleIsInitialized == true) {
          removeDisMoiBubble()
        }

        addNewFloatingDisMoiMessage(x, y)
        promise.resolve("")
      } catch (e: Exception) {
        promise.reject("0", "")
      }
    }
  }

  @ReactMethod
  fun hideFloatingDisMoiBubble(promise: Promise) {
    bubblesManager!!.initialize()
    if (bubbleIsInitialized == true) {
      removeDisMoiBubble()
    }

    promise.resolve("")
  }

  @ReactMethod
  fun hideFloatingDisMoiMessage(promise: Promise) {
    messagesManager!!.initialize()
    removeDisMoiMessage()
    promise.resolve("")
  }

/**
   * Searches for all URLSpans in current text replaces them with our own ClickableSpans
   * forwards clicks to provided function.
   */
  fun TextView.handleUrlClicks(onClicked: ((String) -> Unit)? = null) {
    // create span builder and replaces current text with it
    text = SpannableStringBuilder.valueOf(text).apply {
      // search for all URL spans and replace all spans with our own clickable spans
      this.getSpans(0, length, URLSpan::class.java).forEach {
        // add new clickable span at the same position
        setSpan(
          object : ClickableSpan() {
            override fun onClick(widget: View) {
              onClicked?.invoke(it.url)
            }
          },
          getSpanStart(it),
          getSpanEnd(it),
          Spanned.SPAN_INCLUSIVE_EXCLUSIVE
        )
        // remove old URLSpan
        removeSpan(it)
      }
    }
    // make sure movement method is set
    movementMethod = LinkMovementMethod.getInstance()
  }

  private fun addNewFloatingDisMoiMessage(x: Int, y: Int) {
    if (bubbleIsInitialized == true) {
      removeDisMoiBubble()
    }

    // lateinit var bottomSheetBehavior: BottomSheetBehavior<Message>


    messageDisMoiView = LayoutInflater.from(reactContext).inflate(
      R.layout.highlight_messages, messageDisMoiView, false
    ) as Message

    val carouselView = messageDisMoiView!!.findViewById(
      R.id.carouselView
    ) as CarouselView

    carouselView!!.pageCount = _size

    carouselView!!.setViewListener(viewListener)

    val imageButton = messageDisMoiView!!.findViewById<View>(R.id.close) as ImageButton
    imageButton.setOnClickListener {
      sendEventToReactNative("floating-dismoi-message-press", "")
    }

    messagesManager!!.addDisMoiMessage(messageDisMoiView!!, x, y)
  }

  private fun addNewFloatingDisMoiBubble(x: Int, y: Int, numberOfNotice: String) {
    bubbleDisMoiView = LayoutInflater.from(reactContext).inflate(
      R.layout.bubble, bubbleDisMoiView, false
    ) as Bubble

    var textView: TextView? = bubbleDisMoiView!!.findViewById(R.id.number_of_notice)
    textView!!.text = numberOfNotice

    bubbleDisMoiView!!.setOnBubbleRemoveListener(object : OnBubbleRemoveListener {
      override fun onBubbleRemoved(bubble: Bubble?) {
        bubbleDisMoiView = null
        sendEventToReactNative("floating-dismoi-bubble-remove", "")
      }
    })
    bubbleDisMoiView!!.setOnBubbleClickListener(object : OnBubbleClickListener {
      override fun onBubbleClick(bubble: Bubble?) {
        sendEventToReactNative("floating-dismoi-bubble-press", "")
      }
    })
    bubbleDisMoiView!!.setShouldStickToWall(true)

    bubblesManager!!.addDisMoiBubble(bubbleDisMoiView, x, y)
  }

  private fun removeDisMoiBubble() {
    if (bubbleDisMoiView != null) {
      bubblesManager!!.removeBubble(bubbleDisMoiView)
      bubbleDisMoiView = null
    }
  }

  private fun removeDisMoiMessage() {
    if (messageDisMoiView != null) {
      messagesManager!!.removeDisMoiMessage(messageDisMoiView)
      messageDisMoiView = null
    }
  }

  private fun sendEventToReactNative(eventName: String, params: String) {
    reactContext
      .getJSModule(RCTDeviceEventEmitter::class.java)
      .emit(eventName, params)
  }

  fun String.toSpanned(): Spanned {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      return Html.fromHtml(this, Html.FROM_HTML_MODE_LEGACY)
    } else {
      @Suppress("DEPRECATION")
      return Html.fromHtml(this)
    }
  }

  //set view attributes here
  var viewListener: ViewListener = object : ViewListener {

    override fun setViewForPosition(position: Int): View? {
      val message: ReadableMap? = _notices.getMap(position)

      var disMoiContributorNameMap: ReadableMap? = message!!.getMap("contributor")

      var url: String? = disMoiContributorNameMap!!.getMap("avatar")!!.getMap("normal")!!.getString("url");

      var modified: String? = message!!.getString("modified");

      Log.d("Notification", "created")
      Log.d("Notification", modified.toString());
      Log.d("Notification", "url")
      Log.d("Notification", url.toString());

      var disMoiContributorName: String? = disMoiContributorNameMap.getString("name")

      var disMoiMessage: String? = message.getString("message")

      val customView: View = LayoutInflater.from(reactContext).inflate(R.layout.message, null)

      var textView: TextView? = customView!!.findViewById(R.id.link)

      textView!!.text = disMoiMessage!!.toSpanned()

      var textViewContributorName: TextView? = customView!!.findViewById(R.id.name)

      var textViewDate: TextView? = customView!!.findViewById(R.id.date)

      textViewContributorName!!.text = disMoiContributorName

      textViewDate!!.text = modified

      textView!!.handleUrlClicks { url ->
        sendEventToReactNative("URL_CLICK_LINK", Uri.parse(url).toString())
      }

      return customView
    }
  }
}